design-documents/service-isolation/inter-service-communication.md

```diff
@@ -0,0 +1,301 @@
+### Inter-service communications
+This documents describes how exactly services will be communicating in a distributed setup.
+#### Service declaration
+Services will be declared in Api modules using interfaces. These modules will be installed on all nodes.
+Example:
+```php
+namespace Magento\ServiceBApi;
+
+interface BManagerInterface
+{
+    public function save(BDataInterface $data): BDataInterface;
+    
+    public function delete(string $id): void;
+}
+```
+or for DTOs
+```php
+namespace Magento\ServiceBApi\Data;
+
+interface BDataInterface
+{
+    public function getId(): string;
+  
+    public function getName(): string;
+}
+```
+The API modules will just contain these declarations and config files (like acl.xml, config.xml) describing module
+without describing actual implementation.
+ 
+Even services that are not written using Magento (or even PHP) will be declared in this way for nodes using Magento.
+ 
+#### Publicly available Web APIs
+To enable publicly available APIs (that frontend and 3rd party systems can use) the _front node_ will have to have -Module-Webapi
+or -Module-Graphql modules installed.
+ 
+#### Real and proxy modules
+If a node wants to use services locally then it will just have real modules installed (e.g. "Catalog"), if a node is supposed
+to access services remotely it will have proxy modules installed (e.g. "CatalogProxy").
+ 
+Proxy modules will contain services/DTOs implementations and a di.xml file to declare preferences for proxy implementations to be
+used.
+ 
+For a service such implementation will look like this:
+```php
+namespace Magento\ServiceBProxy;
+
+class BManagerProxy implements Magento\ServiceBApi\BManagerInterface
+{
+    ....
+    
+    public function save(BDataInterface $data): BDataInterface
+    {
+        $promise = $this->invoker->invoker(
```

Comment:
What is the benefit of using promise here?

Reply:
to be able to work with asynchronous operations - the invoker sends requests to other services in an asynchronous manner